src/library/scala/runtime/Statics.java

```diff
@@ -117,4 +122,48 @@ public static int anyHash(Object x) {
 
   /** Used as a marker object to return from PartialFunctions */
   public static final Object pfMarker = new Object();
+
+  // @ForceInline would be nice here.
+  public static void releaseFence() throws Throwable {
+    VM.RELEASE_FENCE.invoke();
+  }
+
+  final static class VM {
+      static final MethodHandle RELEASE_FENCE;
+
+      static {
+          RELEASE_FENCE = mkHandle();
+      }
+
+      private static MethodHandle mkHandle() {
+          MethodHandles.Lookup lookup = MethodHandles.lookup();
+          try {
+              return lookup.findStatic(Class.forName("java.lang.invoke.VarHandle"), "releaseFence", MethodType.methodType(Void.TYPE));
+          } catch (ClassNotFoundException e) {
+              try {
+                  Class<?> unsafeClass = Class.forName("sun.misc.Unsafe");
+                  return lookup.findVirtual(unsafeClass, "storeFence", MethodType.methodType(void.class)).bindTo(findUnsafe(unsafeClass));
+              } catch (NoSuchMethodException | ClassNotFoundException | IllegalAccessException e1) {
+                  ExceptionInInitializerError error = new ExceptionInInitializerError(e1);
+                  error.addSuppressed(e);
+                  throw error;
+              }
+          } catch (NoSuchMethodException | IllegalAccessException e) {
+              throw new ExceptionInInitializerError(e);
+          }
+      }
+
+      private static Object findUnsafe(Class<?> unsafeClass) throws IllegalAccessException {
+          Object found = null;
+          for (Field field : unsafeClass.getDeclaredFields()) {
```

Comment:
I don't have experience with Unsafe, so I googled a bit. https://dzone.com/articles/understanding-sunmiscunsafe suggest that you could create a new Unsafe instance instead of getting the singleton, as the latter doesn't work on Android. Not sure if that's still relevant.

Apart from that, the use of `setAccessible(true)` in List requires basically reflection to be allowed when running any Scala program. Was that the case before? Could security managers be configured in a fine-grained way to allow reflection only for this particular access?

Reply:
Unless things changed without me noticing Android is not really Java8+ so Scala nowadays does not work on it in any case, so most likely this should not be a consideration here. Though one may want to check if I'm up to date about that.

Reply:
Good point re: `setAccessible`. A similar issue was reported with our use of that in structural types: https://github.com/scala/bug/issues/2318#issuecomment-292370451

In that case, we catch the `SecurityException` and fall back to doing nothing.

We could do something similar here, I guess, but it would be good to alert the user somehow that the memory barriers aren't supported in that configuration. For instance, we could throw an exception that says "restart with -Dscala.barrier.noop=true to accept this limitation". Not nice :(

Reply:
This would only affect users on Java 8, of course. On 9+ we just use the public `VarHandle` API.

Reply:
Not very nice, but I think that's our only option. Silently skipping the fence is certainly not a good idea.

Reply:
Looks like the proposal discussed here wasn't implemented in this PR. Should a ticket on http://github.com/scala/scala-dev/ be created to track this?